src/lsm/tree.zig

```diff
@@ -933,18 +929,19 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
             tree.manifest.compact(compact_manifest_callback);
         }
 
+        /// Called after the fourth/last beat.
         fn compact_mutable_table_into_immutable(tree: *Tree) void {
-            // Ensure mutable table can be flushed into immutable table.
             if (tree.table_mutable.count() == 0) return;
             assert(tree.table_immutable.free);
+            assert((tree.compaction_op + 1) % config.lsm_batch_multiple == 0);
 
             // Sort the mutable table values directly into the immutable table's array.
             const values_max = tree.table_immutable.values_max();
             const values = tree.table_mutable.sort_into_values_and_clear(values_max);
             assert(values.ptr == values_max.ptr);
 
-            // Take a manifest snapshot and setup the immutable table with the sorted values.
-            const snapshot_min = tree.compaction_table_immutable.snapshot;
+            const snapshot_min = compaction_snapshot_for_op(tree.compaction_op) + 1;
```

Comment:
Can we explain the plus one with a comment?

If my understanding is correct, it's because:

`// The immutable table should be visible to the next measure.`

However, I think we may have a bug in the plus one—i.e. don't we need to add a full measure's worth of ops?

Reply:
> However, I think we may have a bug in the plus one—i.e. don't we need to add a full measure's worth of ops?

Right now it just needs to be less-than-or-equal to `prefetch_snapshot_max` so that it is read by `lookup_from_memory()` (which we assert on the next line — since `prefetch_snapshot_max` is monotonically increasing, this will always hold). Adding `lsm_batch_multiple` instead of `1` would also work (in that case it would be equal to the `prefetch_snapshot_max`, I think). Or even omit both. My reason for the +1 was to mirror `snapshot_min` of a compaction's output tables (which this sort-of is, if you squint). (The reason we are computing `compaction_snapshot_for_op(tree.compaction_op) + 1` here instead of using `compaction.snapshot + 1` is the `compaction` may not have been run in the last measure, in which case its snapshot is invalid).

:thinking: But it depends how read snapshots will work, which I don't have a good picture of yet.

Reply:
Let's walk and talk through this tomorrow. My mental model is not yet clear enough.